File: Java/JDI/jdi-uitest-mobile/src/main/java/com/epam/jdi/uitests/mobile/appium/elements/common/CheckBox.java
package com.epam.jdi.uitests.mobile.appium.elements.common;
/*
 * Copyright 2004-2016 EPAM Systems
 *
 * This file is part of JDI project.
 *
 * JDI is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * JDI is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; 
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with JDI. If not, see <http://www.gnu.org/licenses/>.
 */


import com.epam.jdi.uitests.core.interfaces.common.ICheckBox;
import com.epam.jdi.uitests.mobile.appium.elements.base.Clickable;
import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;

import java.util.function.Function;

import static com.epam.jdi.uitests.core.settings.JDISettings.exception;

/**
 * Checkbox control implementation
 *
 * @author Alexeenko Yan
 * @author Belousov Andrey
 */
public class CheckBox extends Clickable implements ICheckBox {
    private Function<WebElement, Boolean> isSelected = WebElement::isSelected;
    private Function<WebElement, Boolean> isChecked = el -> el.getAttribute("checked") != null;
    private Function<WebElement, Boolean> isCheckedFunc = el -> {
        if (isSelected.apply(el)) {
            isCheckedFunc = isSelected;
            return true;
        }
        if (isChecked.apply(el)) {
            isCheckedFunc = isChecked;
            return true;
        }
        return false;
    };

    public CheckBox() {
    }

    public CheckBox(By byLocator) {
        super(byLocator);
    }

    public CheckBox(WebElement webElement) {
        super(webElement);
    }

    protected String getValueAction() {
        return isChecked() + "";
    }

    protected void setValueAction(String value) {
        switch (value.toLowerCase()) {
            case "true":
            case "1":
            case "check":
                check();
                break;
            case "false":
            case "0":
            case "uncheck":
                uncheck();
                break;
        }
    }

    protected void checkAction() {
        if (!isCheckedAction())
            clickAction();
        if (!isCheckedAction())
            throw exception("Can't check element. Verify locator for click or isCheckedAction");
    }

    protected void uncheckAction() {
        if (isCheckedAction())
            clickAction();
        if (isCheckedAction())
            throw exception("Can't uncheck element. Verify locator for click or isCheckedAction");
    }

    protected boolean isCheckedAction() {
        return isCheckedFunc.apply(getWebElement());
    }

    public final void check() {
        actions.check(this::checkAction);
    }

    public final void uncheck() {
        actions.uncheck(this::uncheckAction);
    }

    public final boolean isChecked() {
        return actions.isChecked(this::isCheckedAction);
    }

    public final String getValue() {
        return actions.getValue(this::getValueAction);
    }

    public final void setValue(String value) {
        actions.setValue(value, this::setValueAction);
    }
}

File: C#.Net/JDI/Web/Selenium/Elements/Common/CheckBox.cs
using System;
using Epam.JDI.Core.Interfaces.Common;
using Epam.JDI.Web.Selenium.Elements.Base;
using Epam.JDI.Core;
using Epam.JDI.Web.Selenium.Base;
using static Epam.JDI.Core.Settings.JDISettings;

namespace Epam.JDI.Web.Selenium.Elements.Common
{
    public class CheckBox : Clickable, ICheckBox
    {
        protected void CheckAction(WebBaseElement el)
        {
            if (!IsCheckedAction(el))
                ClickAction(el);
            if (!IsCheckedAction(el))
                throw Exception("Can't check element. Verify locator for click or isCheckedAction");
        }
        public void Check()
        {
            Actions.Check(CheckAction);
        }

        protected void UncheckAction(WebBaseElement el)
        {
            if (IsCheckedAction(el))
                ClickAction(el);
            if (IsCheckedAction(el))
                throw Exception("Can't uncheck element. Verify locator for click or isCheckedAction");
        }
        public void Uncheck()
        {
            Actions.Uncheck(UncheckAction);
        }

        protected bool IsCheckedAction(WebBaseElement el)
        {
            return ((CheckBox)el).IsCheckedFunc(el);
        }

        private Func<WebBaseElement, bool> IsCheckedFunc =
            el =>
            {
                var checkbox = (CheckBox)el;
                if (checkbox.IsSelected.Invoke(el))
                    el.WebAvatar.FindImmediately(() => checkbox.IsSelected(el), true);
                if(checkbox.IsCheckedByAttribute.Invoke(el))
                    el.WebAvatar.FindImmediately(() => checkbox.IsCheckedByAttribute(el), true);
                return false;
            };


        private Func<WebBaseElement, bool> IsCheckedByAttribute =
            el => el.WebAvatar.FindImmediately(() => el.WebElement.GetAttribute("checked") != null, false);

        private Func<WebBaseElement, bool> IsSelected =
            el => el.WebAvatar.FindImmediately(() => el.WebElement.Selected, false);

        public bool IsChecked()
        {
            return Actions.IsChecked(IsCheckedAction);
        }
        protected void SetValueAction(WebBaseElement el, string value)
        {
            switch (value.ToLower())
            {
                case "true":
                case "1":
                case "check":
                    ((CheckBox)el).Check();
                    break;
                case "false":
                case "0":
                case "uncheck":
                    ((CheckBox)el).Uncheck();
                    break;
                default:
                    throw Exception($"SetValue not specified correctly {value}, expected: 'true','false','0','1','check','uncheck'");
            }
        }

        public string Value
        {
            get { return Actions.GetValue(GetValueFunc); }
            set { Actions.SetValue(value, SetValueAction); }
        }

        protected Func<WebBaseElement, string> GetValueFunc = el => ((CheckBox) el).IsChecked() + "";

    }
}
